src/System.Threading.Tasks/tests/Task/TaskStatusTest.cs

```diff
@@ -172,11 +172,21 @@ internal void RealRun()
                         {
                             //we may have reach this point too soon, let's keep spinning until the status changes.
                             while (_task.Status == TaskStatus.Running)
-                                ;
+                            {
+                                Task.Delay(1).Wait();
+                            }
+
                             //
-                            // If we're still waiting for children our Status should reflect so
+                            // If we're still waiting for children our Status should reflect so. For this verification, read the
+                            // parent task's status before the child task's status to make the child task's status more recent,
+                            // since the child task may complete during the status reads.
                             //
-                            if (_task.Status != TaskStatus.WaitingForChildrenToComplete)
+                            TaskStatus taskStatus = _task.Status;
+                            Interlocked.MemoryBarrier();
+                            TaskStatus childTaskStatus = _childTask.Status;
+                            if (taskStatus != TaskStatus.WaitingForChildrenToComplete &&
+                                childTaskStatus != TaskStatus.RanToCompletion &&
+                                childTaskStatus != TaskStatus.Faulted)
```

Comment:
The child task is not canceled, but I have changed to use `IsCompleted` as you suggested. I didn't notice the volatile since the tooltip didn't show it :), fixed as well.